File: java/src/main/java/finance/clarian/sdk/ClarianException.java
package finance.clarian.sdk;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Thrown for any non-2xx HTTP response from the Clarian API.
 *
 * <p>The exception message truncates the raw body to 500 characters so logs stay
 * safe when bodies echo request fields (PIX keys, documents). Full body remains
 * available via {@link #getBody()}.
 */
public final class ClarianException extends RuntimeException {

    private final int status;
    private final String code;
    private final String message;
    private final Map<String, Object> meta;
    private final String body;

    public ClarianException(int status, String body) {
        super(buildExceptionMessage(status, body));
        String raw = body == null ? "" : body;
        String code = "";
        String detail = "";
        Map<String, Object> meta = new LinkedHashMap<>();

        if (!raw.isEmpty()) {
            try {
                Object parsed = Json.decode(raw);
                if (parsed instanceof Map<?, ?> map) {
                    Object errVal = map.get("error");
                    if (errVal != null) {
                        code = String.valueOf(errVal);
                    }
                    Object detailVal = map.get("detail");
                    if (detailVal != null) {
                        detail = String.valueOf(detailVal);
                    }
                    for (Map.Entry<?, ?> e : map.entrySet()) {
                        String k = String.valueOf(e.getKey());
                        if ("error".equals(k) || "detail".equals(k)) {
                            continue;
                        }
                        meta.put(k, e.getValue());
                    }
                }
            } catch (RuntimeException ignored) {
                // non-JSON body
            }
        }

        String truncated = raw.length() <= 500 ? raw : raw.substring(0, 500);
        String msg;
        if (!detail.isEmpty()) {
            msg = detail;
        } else if (!code.isEmpty()) {
            msg = code;
        } else {
            msg = truncated;
        }

        this.status = status;
        this.code = code;
        this.message = msg;
        this.meta = Collections.unmodifiableMap(meta);
        this.body = raw;
    }

    private static String buildExceptionMessage(int status, String body) {
        String raw = body == null ? "" : body;
        String code = "";
        if (!raw.isEmpty()) {
            try {
                Object parsed = Json.decode(raw);
                if (parsed instanceof Map<?, ?> map && map.get("error") != null) {
                    code = String.valueOf(map.get("error"));
                }
            } catch (RuntimeException ignored) {
                // keep empty code
            }
        }
        String truncated = raw.length() <= 500 ? raw : raw.substring(0, 500);
        if (!code.isEmpty()) {
            return "HTTP " + status + ": " + code;
        }
        if (!truncated.isEmpty()) {
            return "HTTP " + status + ": " + truncated;
        }
        return "HTTP " + status;
    }

    public int getStatus() {
        return status;
    }

    public String getCode() {
        return code;
    }

    /** Human-readable message (detail, code, or truncated body). */
    public String getErrorMessage() {
        return message;
    }

    /** Remaining JSON fields after error/detail are peeled off (includes hint). */
    public Map<String, Object> getMeta() {
        return meta;
    }

    /** Full raw response body (untruncated). */
    public String getBody() {
        return body;
    }
}

File: php/src/ClarianException.php
<?php

declare(strict_types=1);

namespace Clarian;

/**
 * Thrown for any non-2xx HTTP response from the Clarian API.
 *
 * Exception message truncates the raw body to 500 characters so logs stay safe
 * when bodies echo request fields (PIX keys, documents). Full body remains on
 * {@see $body}.
 *
 * Public fields (also via property access): status, code, message, meta, body.
 * Note: Exception already defines protected $code / $message, so those contract
 * fields are exposed through {@see __get()}.
 */
final class ClarianException extends \RuntimeException
{
    public readonly int $status;

    /** @var array<string, mixed> */
    public readonly array $meta;

    /** Full raw response body (untruncated). */
    public readonly string $body;

    private readonly string $apiCode;
    private readonly string $apiMessage;

    public function __construct(int $status, ?string $body = null)
    {
        $raw = $body ?? '';
        $code = '';
        $detail = '';
        $meta = [];

        if ($raw !== '') {
            try {
                $parsed = json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
                if (is_array($parsed)) {
                    if (array_key_exists('error', $parsed) && $parsed['error'] !== null) {
                        $code = (string) $parsed['error'];
                    }
                    if (array_key_exists('detail', $parsed) && $parsed['detail'] !== null) {
                        $detail = (string) $parsed['detail'];
                    }
                    foreach ($parsed as $k => $v) {
                        if ($k === 'error' || $k === 'detail') {
                            continue;
                        }
                        $meta[(string) $k] = $v;
                    }
                }
            } catch (\JsonException) {
                // non-JSON body
            }
        }

        $truncated = strlen($raw) <= 500 ? $raw : substr($raw, 0, 500);
        if ($detail !== '') {
            $message = $detail;
        } elseif ($code !== '') {
            $message = $code;
        } else {
            $message = $truncated;
        }

        if ($code !== '') {
            $excMsg = "HTTP {$status}: {$code}";
        } elseif ($truncated !== '') {
            $excMsg = "HTTP {$status}: {$truncated}";
        } else {
            $excMsg = "HTTP {$status}";
        }

        parent::__construct($excMsg);
        $this->status = $status;
        $this->apiCode = $code;
        $this->apiMessage = $message;
        $this->meta = $meta;
        $this->body = $raw;
    }

    public function __get(string $name): mixed
    {
        return match ($name) {
            'code' => $this->apiCode,
            'message' => $this->apiMessage,
            default => throw new \Error('Undefined property ClarianException::$' . $name),
        };
    }

    public function __isset(string $name): bool
    {
        return $name === 'code' || $name === 'message';
    }
}
